miden-lib/asm/miden/kernels/tx/account.masm

```diff
@@ -372,16 +389,32 @@ end
 #! - KEY is the key to set.
 #! - OLD_VALUE is the previous value of the item.
 #! - NEW_ROOT is the new root of the map.
-export.set_map_item
-    # TODO: we execute `push.1 drop` before `emit` as decorators are not supported without other
-    #       instructions - see: https://github.com/0xPolygonMiden/miden-vm/issues/1122
-    # emit event to signal that an account storage item is being updated
-    push.1 drop emit.ACCOUNT_STORAGE_SET_MAP_ITEM_EVENT drop
+export.set_map_item.2
+    push.0 drop                                             # TODO: remove line, see miden-vm/#1122
+    emit.ACCOUNT_STORAGE_BEFORE_SET_MAP_ITEM_EVENT
+    # => [index, KEY, NEW_VALUE, OLD_ROOT]
+
+    # duplicate the KEY and the NEW_VALUE to be able to emit an event after an account storage item is 
+    # being updated
+    movdn.12 movupw.2 dupw.2 dupw.2
+    # => [KEY, NEW_VALUE, OLD_ROOT, KEY, NEW_VALUE, index, ...]
 
     # set the NEW_VALUE under KEY in the tree
     # note smt::set expects the stack to be [NEW_VALUE, KEY, OLD_ROOT, ...]
     swapw exec.smt::set
-    # => [OLD_VALUE, NEW_ROOT, ...]
+    # => [OLD_VALUE, NEW_ROOT, KEY, NEW_VALUE, index, ...]
+    
+    # store OLD_VALUE and NEW_ROOT until the end of the procedure 
+    loc_storew.0 dropw loc_storew.1 dropw
+    # => [KEY, NEW_VALUE, index, ...]
+
+    # emit event to signal that an account storage item is being updated
+    movup.8 emit.ACCOUNT_STORAGE_AFTER_SET_MAP_ITEM_EVENT drop dropw dropw
+    # => []
+
+    # load OLD_VALUE and NEW_ROOT on the top of the stack 
+    loc_loadw.1 padw loc_loadw.0
+    # => [OLD_VALUE, NEW_ROOT]
```

Comment:
We don't need to do `dropw dropw` at the end of line 412 because we can overwrite these values with `loc_loadw`s on line 416.

Also, something seems off here: we do two `dropw` and have an empty stack - but then do 2 `loc_loadw` and only one `padw`.

Reply:
I think we indeed can remove `dropw dropw` instructions from the line 412 and use `swapw` instead of `padw` in line 416. 

I used just one `padw` instruction because `loc_loadw` can override only top word on the stack, so I had to pad the stack with an additional space to use the second `loc_loadw`.